### PR TITLE
SA-3040: Add/Remove-JCCloudDirectory Member Functions

### PR DESCRIPTION
**Pull request by gweinjc**
## Issues
* [SA-3040](https://jumpcloud.atlassian.net/browse/SA-3040) - Add/Remove-JCCloudDirectory Member Functions

## What does this solve?
This introduces the following functions:
* Add-JCOffice365Member
* Add-JCGsuiteMember
* Remove-JCOffice365Member
* Remove-JCGsuiteMember

Currently, admins need to utilize the Add/Remove-JCAssociation functions to adjust their memberships to these groups. With the introduction of the Get/Set-JCCloudDirectory functions, it would make it easier on the admins to programatically adjust their CloudDirectory memberships via easy and intuitive functions.

## Is there anything particularly tricky?
N/A

## How should this be tested?
Pester test suites have been created for each of the functions. However we should test to ensure that correct information is being passed into the function, example:

1. `Add-JCOffice365Member -ID <Office365InstanceID> -Username John.Doe -UserID <userID>` | This should throw, only one property should be passed per call (either username OR userid, not both)
2. `Add-JCOffice365Member -ID <Office365InstanceID> -GroupName Office365Users -GroupID <groupID>` | This should throw, only one property should be passed per call (either groupName OR groupID, not both)
3. `Add-JCOffice365Member -ID <Office365InstanceID> -Username John.Doe -GroupName 'Office365Users'` | This should throw. Cannot specify both user and usergroup at the same time

These 3 exceptions should be present for all 4 functions

### CODE REVIEW

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1 [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        [String]$GroupName
+    )
+    begin {
+        Write-Debug 'Verifying JCAPI Key'
+        if ($JCAPIKEY.length -ne 40) {
+            Connect-JCOnline
+        }
+        $resultsArray = [System.Collections.Generic.List[PSObject]]::new()
+        $DirectoryHash = Get-JcSdkDirectory | Where-Object type -EQ 'g_suite' | Select-Object id, name
+        if (($Username -or $UserID) -and ($GroupID -or $GroupName)) {
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:24]: I spent a bit of time thinking about this but I do think this is the only way to get around the many parameters we have here.

Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:24]: In other words I saw this and thought we could do something else but I think this is fine :). Just wanted to call it out to generate conversation.

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1 [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        } elseif ($Username -and $UserID) {
+            throw "Please use either a username or a userID"
+        } elseif ($GroupID -and $GroupName) {
+            throw "Please use either a UserGroup Name or a UserGroup ID"
+        } elseif ($Username -or $UserID) {
+            Write-Debug 'Populating UserHash'
+            $UserHash = Get-DynamicHash -Object User -returnProperties username
+        } elseif ($GroupID -or $GroupName) {
+            Write-Debug 'Populating UserGroupHash'
+            $UserGroupHash = Get-DynamicHash -Object Group -GroupType User -returnProperties name
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:35]: Just wondering about the efficiency here. Technically speaking, if we are going to add a list of users to a cloud directory object. This doesn't feel like the best way for us to be doing this. 

Imagine a list of 10 users we want to add to some Google Workspace Directory:
User_0
User_1
...
User_9

and we write a loop to do this, each user is added with 
`Add-JCGSuiteMember -Name "GoogleDir" -Username "User_0"`

We'd be getting **A LOT** of data somewhat unnecessarily. I know that the DynamicHash functions are quick but it's a lot of data to gather each iteration for just finding one user. I'd really argue that we are better off with the search functions here. i.e.
If Username:
Get-JCUser -username "User_0" -returnProperties username
If UserID:
Get-JCUser -userID "12345" -returnProperties username

We'd still be giving the end user a warning if the username or userID was not found but in this case I'd really argue that getting all users is not performant.

I think we'd want to do the same thing with groups.

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1 [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    process {
+        switch ($PSCmdlet.ParameterSetName) {
+            ByName {
+                Write-Debug "Finding directory by Name"
+                $CloudDirectory = $DirectoryHash | Where-Object name -EQ $Name
+                if (!$CloudDirectory) {
+                    throw "$Name was not found. Please try again"
+                } elseif ($CloudDirectory.count -gt 1) {
+                    throw "Multiple directories with the same name detected, please use the -id parameter to specify which directory to edit"
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:46]: Thanks for this, it's something we would have had to address, I doubt many folks will have this problem but good to catch it before it comes up

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1 [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                }
+            }
+            ByID {
+                Write-Debug "Finding directory by ID"
+                $CloudDirectory = $DirectoryHash | Where-Object Id -EQ $ID
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:51]: I'm less concerned about DirectoryHash but I'm tempted to say that if the customer uses ByID parameter set, there's no need to go get the hash, either they have the correct ID or not. Getting the hash at that stage is unnecessary in my opinion.

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1 [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                if (!$CloudDirectory) {
+                    throw "$ID was not found. Please try again"
+                }
+            }
+        }
+        if ($Username -or $UserID) {
+            if ($Username) {
+                if ($UserHash.Values.username -contains ($Username)) {
+                    $UserID = $UserHash.GetEnumerator().Where({ $_.Value.username -contains ($Username) }).Name
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:60]: Just re-mentioning what I wrote above, I'd argue we should get a user by username here not pull all user data.

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCOffice365Member.ps1 [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        } elseif ($Username -and $UserID) {
+            throw "Please use either a username or a userID"
+        } elseif ($GroupID -and $GroupName) {
+            throw "Please use either a UserGroup Name or a UserGroup ID"
+        } elseif ($Username -or $UserID) {
+            Write-Debug 'Populating UserHash'
+            $UserHash = Get-DynamicHash -Object User -returnProperties username
+        } elseif ($GroupID -or $GroupName) {
+            Write-Debug 'Populating UserGroupHash'
+            $UserGroupHash = Get-DynamicHash -Object Group -GroupType User -returnProperties name
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Add-JCOffice365Member.ps1:35]: Re: issues brought up with Add-JCGsuiteMember, I want to advocate for **not** building a hash for every user/ group

File: PowerShell/JumpCloud Module/Public/Directory/Remove-JCGsuiteMember.ps1 [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        } elseif ($Username -and $UserID) {
+            throw "Please use either a username or a userID"
+        } elseif ($GroupID -and $GroupName) {
+            throw "Please use either a UserGroup Name or a UserGroup ID"
+        } elseif ($Username -or $UserID) {
+            Write-Debug 'Populating UserHash'
+            $UserHash = Get-DynamicHash -Object User -returnProperties username
+        } elseif ($GroupID -or $GroupName) {
+            Write-Debug 'Populating UserGroupHash'
+            $UserGroupHash = Get-DynamicHash -Object Group -GroupType User -returnProperties name
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Remove-JCGsuiteMember.ps1:35]: Re: issues brought up with Add-JCGsuiteMember, I want to advocate for not building a hash for every user/ group

File: PowerShell/JumpCloud Module/Public/Directory/Remove-JCOffice365Member.ps1 [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        } elseif ($Username -and $UserID) {
+            throw "Please use either a username or a userID"
+        } elseif ($GroupID -and $GroupName) {
+            throw "Please use either a UserGroup Name or a UserGroup ID"
+        } elseif ($Username -or $UserID) {
+            Write-Debug 'Populating UserHash'
+            $UserHash = Get-DynamicHash -Object User -returnProperties username
+        } elseif ($GroupID -or $GroupName) {
+            Write-Debug 'Populating UserGroupHash'
+            $UserGroupHash = Get-DynamicHash -Object Group -GroupType User -returnProperties name
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Remove-JCOffice365Member.ps1:35]: Re: issues brought up with Add-JCGsuiteMember, I want to advocate for not building a hash for every user/ group

File: PowerShell/JumpCloud Module/Tests/Public/Directory/Add-JCGsuiteMember.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Added'
+    }
+    It 'Adds a userGroup by ID with directory Name' {
+        $User = Add-JCGSuiteMember -Name $Directories.Name -GroupID $NewGroup.Id
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Added'
+    }
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Tests/Public/Directory/Add-JCGsuiteMember.Tests.ps1:55]: Only thing I want to see in these tests are that the function throws when:
You specify:
Name & ID params together
Username & UserID together
Groupname & GroupID together

Items do not exist:
Name OR ID where the Name OR ID does not exist
Username OR UserID where the Username OR UserID does not exist
GroupName OR GroupID where the GroupName OR GroupID does not exist

File: PowerShell/JumpCloud Module/Tests/Public/Directory/Add-JCOffice365Member.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Added'
+    }
+    It 'Adds a userGroup by ID with directory Name' {
+        $User = Add-JCOffice365Member -Name $Directories.Name -GroupID $NewGroup.Id
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Added'
+    }
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Tests/Public/Directory/Add-JCOffice365Member.Tests.ps1:55]: Re: Tests for Add-JCGsuiteMember
Tests for params should throw
Tests for items that do not exist should throw

File: PowerShell/JumpCloud Module/Tests/Public/Directory/Remove-JCGsuiteMember.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Removed'
+    }
+    It 'Removes a userGroup by ID with directory Name' {
+        $User = Remove-JCGsuiteMember -Name $Directories.Name -GroupID $NewGroup.Id
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Removed'
+    }
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Tests/Public/Directory/Remove-JCGsuiteMember.Tests.ps1:58]: Re: Tests for Add-JCGsuiteMember
Tests for params should throw
Tests for items that do not exist should throw

File: PowerShell/JumpCloud Module/Tests/Public/Directory/Remove-JCOffice365Member.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Removed'
+    }
+    It 'Removes a userGroup by ID with directory Name' {
+        $User = Remove-JCOffice365Member -Name $Directories.Name -GroupID $NewGroup.Id
+        $User.DirectoryName | Should -Be $Directories.Name
+        $User.GroupID | Should -Be $NewGroup.ID
+        $User.Status | Should -Be 'Removed'
+    }
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Tests/Public/Directory/Remove-JCOffice365Member.Tests.ps1:58]: Re: Tests for Add-JCGsuiteMember
Tests for params should throw
Tests for items that do not exist should throw

**Review by jworkmanjc — CHANGES_REQUESTED** [review]
Overall I think this is really good, I think we do need to make a change in the data that we are gathering in each function. I'm okay to debate it and come up with a team solution if you feel one way or another but I'd argue we shouldn't be building hash objects for users and usergroups.

File: PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1 (hunk 2 above)
Comment (kmaranionjc) [reply PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1:35]: +1 on Joe's point, we can maybe be more efficient by not getting all user/group objects if we are only pulling one or two data.

**Review by kmaranionjc — CHANGES_REQUESTED** [review]
Nice work overall, like Joe stated on his comments, Get-DynamicHash is inefficient for small pulls but something like Search should do.

File: PowerShell/JumpCloud Module/Docs/Set-JCSettingsFile.md [hunk 13]
```diff
@@ -14,7 +14,7 @@ Updates the JumpCloud Module Settings File
 ## SYNTAX
 
 ```
-Set-JCSettingsFile [-parallelOverride <PSObject>] [-moduleBannerMessageCount <PSObject>] [<CommonParameters>]
+Set-JCSettingsFile [-moduleBannerMessageCount <PSObject>] [-parallelOverride <PSObject>] [<CommonParameters>]
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Set-JCSettingsFile.md:17]: I think if we start to add positions to our parameters, it'll prevent these help files from being changed. I can get a card for that.

File: PowerShell/JumpCloud Module/Public/Directory/Remove-JCGsuiteMember.ps1 [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        $DirectoryHash = Get-JcSdkDirectory | Where-Object type -EQ 'g_suite' | Select-Object id, name
+        if (($Username -or $UserID) -and ($GroupID -or $GroupName)) {
+            throw "Please use one type of association per call"
+        } elseif ($Username -and $UserID) {
+            throw "Please use either a username or a userID"
+        } elseif ($GroupID -and $GroupName) {
+            throw "Please use either a UserGroup Name or a UserGroup ID"
+        } elseif ($Username -or $UserID) {
+            Write-Debug 'Populating UserHash'
+            $UserHash = Get-DynamicHash -Object User -returnProperties username
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Directory/Remove-JCGsuiteMember.ps1:32]: Still using dynamicHash for user/ group here

**Review by jworkmanjc — CHANGES_REQUESTED** [review]
Thanks for all the changes, just one last function (`Remove-JCGsuiteMember`) needs that re-work for user/ group hash but otherwise this looks correct. Thanks for the revision

**Review by kmaranionjc — APPROVED** [review]
Thanks for the work. Tests are good on my end.

**Review by jworkmanjc — APPROVED** [review]
Thanks for the changes! Everything looks great!

**Pull request merged**